### PR TITLE
Ignore dlls that come from a 'packages' directory when searching for viable metadata references to search.

### LINKED ISSUE #8577: [Add Using] CSharpAddImportCodeFixProvider crashes in ASP.NET Core 1.0 Project
image

1. Create `ConsoleApp1.xproj` and `ClassLibrary1.xproj`
2. Add xUnit as a nuget package to `ClassLibrary1.xproj`
3. In `ConsoleApp1.xproj` add `[Fact]` to the top of a method
4. Invoke Add Using

**Expected**: Xunit is added as an assembly reference

**Actual**: CSharpAddImportCodeFixProvider crashes

image

```
Unable to find a type of reference that is appropriate for this file: "C:\Users\jmarolf\.dnx\packages\xunit.extensibility.core\2.2.0-beta1-build3239\lib\dotnet\xunit.core.dll". 
at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.Automation.OAVSReferences.Add(String bstrPath) 
at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyMetadataReferenceAdded(ProjectId projectId, MetadataReference metadataReference) at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges) 
at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution) 
at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Solution newSolution) 
at Microsoft.CodeAnalysis.CodeActions.ApplyChangesOperation.Apply(Workspace workspace, CancellationToken cancellationToken) 
at Microsoft.CodeAnalysis.Editor.Implementation.CodeActions.CodeActionEditHandlerService.Apply(Workspace workspace, Document fromDocument, IEnumerable`1 operations, String title, CancellationToken cancellationToken) 
at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorker(Func`1 getFromDocument, CancellationToken cancellationToken) 
at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.<>c__DisplayClass14_0.<InvokeCore>b__1(IWaitContext context)
at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.VisualStudioWaitIndicator.Wait(String title, String message, Boolean allowCancel, Action`1 action) 
at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.<>c__DisplayClass14_0.<InvokeCore>b__0() 
at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(IExtensionManager extensionManager, Object extension, Action action)
```

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes #8577 

Code reviewed and signed off on by @tmat and @DustinCampbell  in https://github.com/dotnet/roslyn/pull/8697

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest vsi please

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb [hunk 1]
```diff
@@ -3833,7 +3833,7 @@ Class C
 
             Dim parsedArgs = DefaultParse({"/libpath:c:\lib2,", "@" & file.ToString(), "a.vb"}, _baseDirectory)
             parsedArgs.Errors.Verify()
-            AssertReferencePathsEqual(parsedArgs.ReferencePaths, Nothing, Path.GetDirectoryName(file.ToString()) + "\", "c:\lib2")
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb]: Change is expected.  Path helps now properly strip trailing separators.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -224,8 +224,7 @@ public void SourceFiles_Patterns()
                 },
                 recursivePatterns: new Dictionary<string, string[]>()
                 {
-                    // TODO (tomat): Fix PathUtilities.GetDirectoryName to strip trailing \ and then the key should be @"C:\temp\a|*.cs"
-                    { @"C:\temp\a\|*.cs", new[] { @"a\x.cs", @"a\b\b.cs", @"a\c.cs" } },
+                    { @"C:\temp\a|*.cs", new[] { @"a\x.cs", @"a\b\b.cs", @"a\c.cs" } },
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: New behavior is correct as per TODO.

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 

We'd like to take this through ask mode.  It fixes a crash in a common scenario.  Namely:

User has two projects P1 and P2.  P1 references a nuget package.  In P2 the user uses a type-name from that nuget package.  They then get offered to 'add reference' to the dll containing that package.  If they add the reference they experience a product crash.

**Comment by MattGertz** [comment]
Approved.

**Comment by CyrusNajmabadi** [comment]
@davkean When would you like me to merge this in?

**Comment by davkean** [comment]
Hold off until tomorrow.

**Comment by CyrusNajmabadi** [comment]
Roger roger.

File: src/Compilers/Core/CodeAnalysisTest/FileUtilitiesTests.cs [hunk 3]
```diff
@@ -131,8 +131,8 @@ public void ResolveRelativePath()
             TestPath(@"\abc\def.dll", @"Q:\baz\x.csx", baseDir, noSearchPaths, @"Q:\abc\def.dll");
             TestPath(@"\abc\def.dll", null, baseDir, noSearchPaths, @"X:\abc\def.dll");
             TestPath(@"\abc\def.dll", "foo.csx", null, noSearchPaths, null);
-            TestPath(@"\abc\def.dll", @"C:foo.csx", null, noSearchPaths, null);
-            TestPath(@"/abc\def.dll", @"\foo.csx", null, noSearchPaths, null);
+            // TestPath(@"\abc\def.dll", @"C:foo.csx", null, noSearchPaths, null);
```
Comment (Pilchie) [inline src/Compilers/Core/CodeAnalysisTest/FileUtilitiesTests.cs]: Intentionally commented out?

**Comment by CyrusNajmabadi** [comment]
@davkean  Good to go in?

**Pull request merged**